### PR TITLE
add-nested-field-dropdown for complex-list

### LINKED ISSUE #301: Nested add-field-dropdown for complex-list items
### UI implementation with dummy logic

https://github.com/harunurhan/ng2-json-editor/tree/add-nested-field (last 3 commits)

### How it looks:

screen shot 2017-05-18 at 16 02 44

### Some challenges

- Hiding all other deeper `add-field-dropdown` for items of `complex-list`
- Building the data structure (to be binded to dropdown) from keys (`keys-store.service`) and schema that is also update when keys are changed.
- Handling the field addition action for `table-list` inside `complex-list`

### PR DESCRIPTION
**Pull request by harunurhan**
* Adds add-nested-field-dropdown for complex-list so that keys can be
added to deeper level objects and table-lists. (Fixes #301)

### CODE REVIEW

**Comment by harunurhan** [comment]
not done yet, but ok to check @jmartinm

File: src/add-field-dropdown/add-nested-field-dropdown.component.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with ng2-json-editor; if not, write to the Free Software Foundation, Inc.,
+ * 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
+ * In applying this license, CERN does not
+ * waive the privileges and immunities granted to it by virtue of its status
+ * as an Intergovernmental Organization or submit itself to any jurisdiction.
+*/
+
+import { Component, Input, Output, EventEmitter, ElementRef, OnInit } from '@angular/core';
```
Comment (jmartinm) [inline src/add-field-dropdown/add-nested-field-dropdown.component.ts]: Unused imports

File: src/add-field-dropdown/add-nested-field-dropdown.component.ts [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    return schemaKeys.subtract(keys);
+  }
+
+  onFieldSelect(path: string, key: string) {
+    let schema = this.jsonSchemaService.forPathString(path);
+    if (schema.componentType === 'table-list') {
+      schema = schema.items;
+    }
+    let newKeyPath = this.keysStoreService.addKey(path, key, schema);
+    if (this.keysStoreService.keysMap[newKeyPath]) {
```
Comment (jmartinm) [inline src/add-field-dropdown/add-nested-field-dropdown.component.ts]: To check if it is necessary

File: src/add-field-dropdown/add-nested-field-dropdown.component.html [hunk 3]
```diff
@@ -0,0 +1,14 @@
+<div class="btn-group add-field-dropdown-container" dropdown keyboardNav="true">
+  <button type="button" class="btn-add-field-dropdown" dropdownToggle>
+    <i class="fa fa-plus"></i> <i class="fa fa-caret-down"></i>
+  </button>
+  <ul class="dropdown-menu" *dropdownMenu>
+    <div *ngFor="let path of nestedKeysMap | keys; first as isFirst">
+      <li *ngIf="!isFirst" class="divider dropdown-divider"></li>
+      <li *ngIf="!isFirst" class="dropdown-header title">{{path | lastPathElement}}</li>
+      <li *ngFor="let key of difference(path)">
```
Comment (jmartinm) [inline src/add-field-dropdown/add-nested-field-dropdown.component.html]: Does not exist.

File: src/complex-list-field/complex-list-field.component.html [hunk 4]
```diff
@@ -49,8 +49,15 @@
       </span>
       <span *ngIf="item.isEditFormVisible">
         <object-field [value]="values.get(item.index)" [schema]="schema.items" [path]="getPathForChild(item.index)">
-          <list-action-group (onMove)="moveElement(item.index, $event)" (onDelete)="deleteElement(item.index)" [canMove]="schema.sortable"></list-action-group>
         </object-field>
+        <div class="row element-button-container">
+          <div class="col-md-6 left">
+            <add-nested-field-dropdown [pathString]="pathString + '/' + item.index" [schema]="schema.items"></add-nested-field-dropdown>
```
Comment (jmartinm) [inline src/complex-list-field/complex-list-field.component.html]: use `getPathForChild`

File: src/complex-list-field/complex-list-field.component.html [hunk 5]
```diff
@@ -49,8 +49,15 @@
...
+          </div>
+          <div class="col-md-6 right">
```
Comment (jmartinm) [inline src/complex-list-field/complex-list-field.component.html]: Maybe you can use the bootstrap class `pull-right`

File: src/shared/pipes/keys-pipe.ts [hunk 6]
```diff
@@ -0,0 +1,38 @@
+/*
+ * This file is part of ng2-json-editor.
+ * Copyright (C) 2016 CERN.
```
Comment (jmartinm) [inline src/shared/pipes/keys-pipe.ts]: 2017

File: src/shared/pipes/last-path-element.pipe.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+ * In applying this license, CERN does not
+ * waive the privileges and immunities granted to it by virtue of its status
+ * as an Intergovernmental Organization or submit itself to any jurisdiction.
+*/
+
+import { Pipe, PipeTransform } from '@angular/core';
+import { PathUtilService } from '../services';
+
+/**
+ * It returns array of keys which aren't present in given object
```
Comment (jmartinm) [inline src/shared/pipes/last-path-element.pipe.ts]: Leftover comment

File: src/shared/pipes/last-path-element.pipe.ts [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+ */
+@Pipe({
+  name: 'lastPathElement'
+})
+export class LastPathElementPipe implements PipeTransform {
+
+  constructor(private pathPathUtilService: PathUtilService) { }
```
Comment (jmartinm) [inline src/shared/pipes/last-path-element.pipe.ts]: `pathUtilService`

File: src/add-field-dropdown/add-nested-field-dropdown.component.ts (hunk 2 above)
Comment (harunurhan) [inline src/add-field-dropdown/add-nested-field-dropdown.component.ts]: I checked and it's for now :)

File: src/complex-list-field/complex-list-field.component.html (hunk 5 above)
Comment (harunurhan) [inline src/complex-list-field/complex-list-field.component.html]: didn't work as expected

File: src/complex-list-field/complex-list-field.component.html (hunk 4 above)
Comment (harunurhan) [inline src/complex-list-field/complex-list-field.component.html]: it returns array, but we need string here so left it like this.

**Pull request merged**